docs/design/event-streams.md

```diff
@@ -0,0 +1,246 @@
+    velocity: Float,
+}
+
+/// An example error emitted when the client is throttled
+/// and should terminate the event stream.
+@error("client")
+@retryable(throttling: true)
+structure ThrottlingError {}
+```
+
+
+### Event Stream Type Representation
+
+The members of an operation input or output that target a stream will be represented using by an asynchronous [Flow](https://kotlinlang.org/docs/reference/coroutines/flow.html) 
+from the `kotlinx-coroutines-core` library. `Flow` is a natural fit for representing asynchronous streams. 
+
+`Flow` was chosen for pagination and already in use as part of our public API contract. Any alternative to this would require a custom but similar type that doesn't play well with
+the rest of the coroutine ecosystem. There is also prior art for representing streaming requests and responses, see [gRPC Kotlin](https://github.com/grpc/grpc-kotlin).
+
+The following types and service would be generated. 
+
+NOTE: only the input and output types are shown, the other structures or unions in the model would be generated as described in [Kotlin Smithy Design](kotlin-smithy-sdk.md).
+
+#### Input Event Streams
+
+
+```kt
+class PublishMessagesRequest private constructor(builder: Builder){
+    val room: String? = builder.room
+    val messages: Flow<PublishEvents>? = builder.messages
+
+    ...
+
+    public class Builder {
+        var room: String? = null
+        var messages: Flow<PublishEvents>? = null
+        fun build(): PublishMessagesRequest = PublishMessagesRequest(this)
+    }
+}
+
+```
+
+#### Output Event Streams
+
+Output event streams would be modeled the same way as input streams. The response object would have a `Flow<T>` field that represents the response stream.
+
+```kt
+class SubscribeToMovementsResponse private constructor(builder: Builder){
+    val movements: Flow<PublishEvents>? = builder.movements
+
+    ...
+
+    public class Builder {
+        var movements: Flow<MovementEvents>? = null
+        fun build(): SubscribeToMovementsResponse = SubscribeToMovementsResponse(this)
+    }
+}
+
+```
+
+
+Modeling the event stream as a field of the request or response allows for [initial messages](https://awslabs.github.io/smithy/1.0/spec/core/stream-traits.html#initial-messages) 
+to be implemented. If we directly returned or took a `Flow<T>` as the input or output type we would not be able to represent the initial request or response fields when present.
+
+
+### **Service and Usage**
+
+NOTE: There are types and internal details here not important to the design of how customers will interact with 
+streaming requests/responses (e.g. serialization/deserialization). 
+Those details are subject to change and not part of this design document. The focus here should be on the way 
+streaming is exposed to a customer.
+
+
+The signatures generated match that of binary streaming requests and responses. Notably that output streams take a lambda instead of returning the response directly.
```

Comment:
suggestion

It might be nice to link to the design doc in which we discuss our similar pattern for handling binary data in response blocks.